astropy/coordinates/solar_system.py

```diff
@@ -211,27 +215,33 @@ def get_body_barycentric(body, time, ephemeris=None):
         body = body.lower()
         earth_pv_helio, earth_pv_bary = erfa.epv00(jd1, jd2)
         if body == 'earth':
-            cartesian_position_body = earth_pv_bary[..., 0, :]
+            body_pv_bary = earth_pv_bary
+
         elif body == 'moon':
-            cartesian_position_body = calc_moon(time).cartesian.xyz.to(u.au).value
-            cartesian_position_body = np.rollaxis(cartesian_position_body, 0,
-                                                  cartesian_position_body.ndim)
+            if get_velocity:
+                raise KeyError("the Moon's velocity cannot be calculated with "
+                               "the '{0}' ephemeris.".format(ephemeris))
+            return calc_moon(time).cartesian
+
         else:
-            sun_bary = earth_pv_bary[..., 0, :] - earth_pv_helio[..., 0, :]
+            sun_pv_bary = earth_pv_bary - earth_pv_helio
             if body == 'sun':
-                cartesian_position_body = sun_bary
+                body_pv_bary = sun_pv_bary
             else:
                 try:
                     body_index = PLAN94_BODY_NAME_TO_PLANET_INDEX[body]
                 except KeyError:
-                    raise KeyError("{0}'s position cannot be calculated with "
-                                   "the '{1}' ephemeris."
+                    raise KeyError("{0}'s position and velocity cannot be "
+                                   "calculated with the '{1}' ephemeris."
                                    .format(body, ephemeris))
                 body_pv_helio = erfa.plan94(jd1, jd2, body_index)
-                cartesian_position_body = body_pv_helio[..., 0, :] + sun_bary
+                body_pv_bary = body_pv_helio + sun_pv_bary
 
-        barycen_to_body_vector = u.Quantity(
-            np.rollaxis(cartesian_position_body, -1, 0), u.au)
+        body_posvel_bary = np.rollaxis(body_pv_bary, -1, 0)
```

Comment:
Is there a reason for using a different array here? Can't you just use body_pv_bary throughout?